src/network/nm/utils.cpp

```diff
@@ -0,0 +1,341 @@
+				}
+			}
+			return false;
+		}
+		break;
+	case NMWirelessSecurityType::Wpa2Psk:
+		if (!(caps & NMWirelessCapabilities::Rsn)) {
+			return false;
+		}
+		if (haveAp) {
+			if (adhoc) {
+				if (!(caps & NMWirelessCapabilities::IbssRsn)) {
+					return false;
+				}
+				if (apRsn & NM80211ApSecurityFlags::PairCcmp && caps & NMWirelessCapabilities::CipherCcmp) {
+					return true;
+				}
+			} else {
+				if (apRsn & NM80211ApSecurityFlags::KeyMgmtPsk) {
+					if (apRsn & NM80211ApSecurityFlags::PairTkip && caps & NMWirelessCapabilities::CipherTkip)
+					{
+						return true;
+					}
+					if (apRsn & NM80211ApSecurityFlags::PairCcmp && caps & NMWirelessCapabilities::CipherCcmp)
+					{
+						return true;
+					}
+				}
+			}
+			return false;
+		}
+		break;
+	case NMWirelessSecurityType::WpaEap:
+		if (adhoc) {
+			return false;
+		}
+		if (!(caps & NMWirelessCapabilities::Wpa)) {
+			return false;
+		}
+		if (haveAp) {
+			if (!(apWpa & NM80211ApSecurityFlags::KeyMgmt8021x)) {
+				return false;
+			}
+			// Ensure at least one WPA cipher is supported
+			if (!deviceSupportsApCiphers(caps, apWpa, NMWirelessSecurityType::WpaEap)) {
+				return false;
+			}
+		}
+		break;
+	case NMWirelessSecurityType::Wpa2Eap:
+		if (adhoc) {
+			return false;
+		}
+		if (!(caps & NMWirelessCapabilities::Rsn)) {
+			return false;
+		}
+		if (haveAp) {
+			if (!(apRsn & NM80211ApSecurityFlags::KeyMgmt8021x)) {
+				return false;
+			}
+			// Ensure at least one WPA cipher is supported
+			if (!deviceSupportsApCiphers(caps, apRsn, NMWirelessSecurityType::Wpa2Eap)) {
+				return false;
+			}
+		}
+		break;
+	case NMWirelessSecurityType::Sae:
+		if (!(caps & NMWirelessCapabilities::Rsn)) {
+			return false;
+		}
+		if (haveAp) {
+			if (adhoc) {
+				if (!(caps & NMWirelessCapabilities::IbssRsn)) {
+					return false;
+				}
+				if (apRsn & NM80211ApSecurityFlags::PairCcmp && caps & NMWirelessCapabilities::CipherCcmp) {
+					return true;
+				}
+			} else {
+				if (apRsn & NM80211ApSecurityFlags::KeyMgmtSae) {
+					if (apRsn & NM80211ApSecurityFlags::PairTkip && caps & NMWirelessCapabilities::CipherTkip)
+					{
+						return true;
+					}
+					if (apRsn & NM80211ApSecurityFlags::PairCcmp && caps & NMWirelessCapabilities::CipherCcmp)
+					{
+						return true;
+					}
+				}
+			}
+			return false;
+		}
+		break;
+	case NMWirelessSecurityType::Owe:
+		if (adhoc) {
+			return false;
+		}
+		if (!(caps & NMWirelessCapabilities::Rsn)) {
+			return false;
+		}
+		if (haveAp) {
+			if (!(apRsn & NM80211ApSecurityFlags::KeyMgmtOwe)
+			    && !(apRsn & NM80211ApSecurityFlags::KeyMgmtOweTm))
+			{
+				return false;
+			}
+		}
+		break;
+	case NMWirelessSecurityType::Wpa3SuiteB192:
+		if (adhoc) {
+			return false;
+		}
+		if (!(caps & NMWirelessCapabilities::Rsn)) {
+			return false;
+		}
+		if (haveAp && !(apRsn & NM80211ApSecurityFlags::KeyMgmtEapSuiteB192)) {
+			return false;
+		}
+		break;
+	default: good = false; break;
+	}
+
+	return good;
+}
```

Comment:
I'm assuming this is copied directly from the NM source, but I think it would be beneficial to rewrite it in a way that is intended to be read. At least removing braces for single-line ifs would go a long way.